File: nrooms.f
C Copyright (c) 1980, InfoCom Computers and Communications, Cambridge MA 02142
C All rights reserved, commercial usage strictly prohibited.
C Written by R. M. Supnik.
C Revisions Copyright (c) 2021, Darth Spectra (Lydia Marie Williamson).
C
C RAPPL2- SPECIAL PURPOSE ROOM ROUTINES, PART 2
	LOGICAL FUNCTION RAPPL2(RI)
	IMPLICIT INTEGER (A-Z)
	LOGICAL QOPEN,QHERE
	include 'parser.h'
	include 'gamestat.h'
	include 'state.h'
	include 'io.h'
	include 'rooms.h'
	include 'rflag.h'
	include 'rindex.h'
	include 'objects.h'
	include 'oflags.h'
	include 'oindex.h'
	include 'xsrch.h'
	include 'clock.h'
	include 'advers.h'
	include 'verbs.h'
	include 'flags.h'
C
C In-line function
	QOPEN(R)=iand(OFLAG2(R),OPENBT).NE.0
C
C Initialized data
	DATA NEWRMS/38/
C RAPPL2, PAGE 2
C
	RAPPL2=.TRUE.
	GO TO (38000,39000,40000,41000,42000,43000,44000,
     &	45000,46000,47000,48000,49000,50000,
     &	51000,52000,53000,54000,55000,56000,
     &	57000,58000,59000,60000),
     &	(RI-NEWRMS+1)
	CALL BUG(70,RI)
	RETURN
C
C R38--	MIRROR D ROOM
C
38000	IF(PRSA.EQ.LOOKW) CALL LOOKTO(FDOOR,MRG,0,682,681)
	RETURN
C
C R39--	MIRROR G ROOM
C
39000	IF(PRSA.EQ.WALKIW) CALL JIGSUP(685)
	RETURN
C
C R40--	MIRROR C ROOM
C
40000	IF(PRSA.EQ.LOOKW) CALL LOOKTO(MRG,MRB,683,0,681)
	RETURN
C
C R41--	MIRROR B ROOM
C
41000	IF(PRSA.EQ.LOOKW) CALL LOOKTO(MRC,MRA,0,0,681)
	RETURN
C
C R42--	MIRROR A ROOM
C
42000	IF(PRSA.EQ.LOOKW) CALL LOOKTO(MRB,0,0,684,681)
	RETURN
C RAPPL2, PAGE 3
C
C R43--	MIRROR C EAST/WEST
C
43000	IF(PRSA.EQ.LOOKW) CALL EWTELL(HERE,683)
	RETURN
C
C R44--	MIRROR B EAST/WEST
C
44000	IF(PRSA.EQ.LOOKW) CALL EWTELL(HERE,686)
	RETURN
C
C R45--	MIRROR A EAST/WEST
C
45000	IF(PRSA.EQ.LOOKW) CALL EWTELL(HERE,687)
	RETURN
C
C R46--	INSIDE MIRROR
C
46000	IF(PRSA.NE.LOOKW) RETURN
C						!LOOK?
	CALL RSPEAK(688)
C						!DESCRIBE
C
C NOW DESCRIBE POLE STATE.
C
C CASES 1,2--	MDIR=270 & MLOC=MRB, POLE IS UP OR IN HOLE
C CASES 3,4--	MDIR=0 V MDIR=180, POLE IS UP OR IN CHANNEL
C CASE 5--	POLE IS UP
C
	I=689
C						!ASSUME CASE 5.
	IF((MDIR.EQ.270).AND.(MLOC.EQ.MRB))
     &	I=690+MIN0(POLEUF,1)
	IF(MOD(MDIR,180).EQ.0)
     &	I=692+MIN0(POLEUF,1)
	CALL RSPEAK(I)
C						!DESCRIBE POLE.
	CALL RSPSUB(694,695+(MDIR/45))
C						!DESCRIBE ARROW.
	RETURN
C RAPPL2, PAGE 4
C
C R47--	MIRROR EYE ROOM
C
47000	IF(PRSA.NE.LOOKW) RETURN
C						!LOOK?
	I=704
C						!ASSUME BEAM STOP.
	DO 47100 J=1,OLNT
	  IF(QHERE(J,HERE).AND.(J.NE.RBEAM)) GO TO 47200
47100	CONTINUE
	I=703
47200	CALL RSPSUB(I,ODESC2(J))
C						!DESCRIBE BEAM.
	CALL LOOKTO(MRA,0,0,0,0)
C						!LOOK NORTH.
	RETURN
C
C R48--	INSIDE CRYPT
C
48000	IF(PRSA.NE.LOOKW) RETURN
C						!LOOK?
	I=46
C						!CRYPT IS OPEN/CLOSED.
	IF(QOPEN(TOMB)) I=12
	CALL RSPSUB(705,I)
	RETURN
C
C R49--	SOUTH CORRIDOR
C
49000	IF(PRSA.NE.LOOKW) RETURN
C						!LOOK?
	CALL RSPEAK(706)
C						!DESCRIBE.
	I=46
C						!ODOOR IS OPEN/CLOSED.
	IF(QOPEN(ODOOR)) I=12
	IF(LCELL.EQ.4) CALL RSPSUB(707,I)
C						!DESCRIBE ODOOR IF THERE.
	RETURN
C
C R50--	BEHIND DOOR
C
50000	IF(PRSA.NE.WALKIW) GO TO 50100
C						!WALK IN?
	CFLAG(CEVFOL)=.TRUE.
C						!MASTER FOLLOWS.
	CTICK(CEVFOL)=-1
	RETURN
C
50100	IF(PRSA.NE.LOOKW) RETURN
C						!LOOK?
	I=46
C						!QDOOR IS OPEN/CLOSED.
	IF(QOPEN(QDOOR)) I=12
	CALL RSPSUB(708,I)
	RETURN
C RAPPL2, PAGE 5
C
C R51--	FRONT DOOR
C
51000	IF(PRSA.EQ.WALKIW) CTICK(CEVFOL)=0
C						!IF EXITS, KILL FOLLOW.
	IF(PRSA.NE.LOOKW) RETURN
C						!LOOK?
	CALL LOOKTO(0,MRD,709,0,0)
C						!DESCRIBE SOUTH.
	I=46
C						!PANEL IS OPEN/CLOSED.
	IF(INQSTF) I=12
C						!OPEN IF INQ STARTED.
	J=46
C						!QDOOR IS OPEN/CLOSED.
	IF(QOPEN(QDOOR)) J=12
	CALL RSPSB2(710,I,J)
	RETURN
C
C R52--	NORTH CORRIDOR
C
52000	IF(PRSA.NE.LOOKW) RETURN
C						!LOOK?
	I=46
	IF(QOPEN(CDOOR)) I=12
C						!CDOOR IS OPEN/CLOSED.
	CALL RSPSUB(711,I)
	RETURN
C
C R53--	PARAPET
C
53000	IF(PRSA.EQ.LOOKW) CALL RSPSUB(712,712+PNUMB)
	RETURN
C
C R54--	CELL
C
54000	IF(PRSA.NE.LOOKW) RETURN
C						!LOOK?
	I=721
C						!CDOOR IS OPEN/CLOSED.
	IF(QOPEN(CDOOR)) I=722
	CALL RSPEAK(I)
	I=46
C						!ODOOR IS OPEN/CLOSED.
	IF(QOPEN(ODOOR)) I=12
	IF(LCELL.EQ.4) CALL RSPSUB(723,I)
C						!DESCRIBE.
	RETURN
C
C R55--	PRISON CELL
C
55000	IF(PRSA.EQ.LOOKW) CALL RSPEAK(724)
C						!LOOK?
	RETURN
C
C R56--	NIRVANA CELL
C
56000	IF(PRSA.NE.LOOKW) RETURN
C						!LOOK?
	I=46
C						!ODOOR IS OPEN/CLOSED.
	IF(QOPEN(ODOOR)) I=12
	CALL RSPSUB(725,I)
	RETURN
C RAPPL2, PAGE 6
C
C R57--	NIRVANA AND END OF GAME
C
57000	IF(PRSA.NE.WALKIW) RETURN
C						!WALKIN?
	CALL RSPEAK(726)
	CALL SCORE(.FALSE.)
C moved to exit routine	CLOSE(DBCH)
	CALL EXIT
C
C R58--	TOMB ROOM
C
58000	IF(PRSA.NE.LOOKW) RETURN
C						!LOOK?
	I=46
C						!TOMB IS OPEN/CLOSED.
	IF(QOPEN(TOMB)) I=12
	CALL RSPSUB(792,I)
	RETURN
C
C R59--	PUZZLE SIDE ROOM
C
59000	IF(PRSA.NE.LOOKW) RETURN
C						!LOOK?
	I=861
C						!ASSUME DOOR CLOSED.
	IF(CPOUTF) I=862
C						!OPEN?
	CALL RSPEAK(I)
C						!DESCRIBE.
	RETURN
C
C R60--	PUZZLE ROOM
C
60000	IF(PRSA.NE.LOOKW) RETURN
C						!LOOK?
	IF(CPUSHF) GO TO 60100
C						!STARTED PUZZLE?
	CALL RSPEAK(868)
C						!NO, DESCRIBE.
	IF(iand(OFLAG2(WARNI),TCHBT).NE.0) CALL RSPEAK(869)
	RETURN
C
60100	CALL CPINFO(880,CPHERE)
C						!DESCRIBE ROOM.
	RETURN
	END
C
C LOOKTO--	DESCRIBE VIEW IN MIRROR HALLWAY
	SUBROUTINE LOOKTO(NRM,SRM,NT,ST,HT)
	IMPLICIT INTEGER(A-Z)
	include 'gamestat.h'
	include 'flags.h'
C LOOKTO, PAGE 2
C
	CALL RSPEAK(HT)
C						!DESCRIBE HALL.
	CALL RSPEAK(NT)
C						!DESCRIBE NORTH VIEW.
	CALL RSPEAK(ST)
C						!DESCRIBE SOUTH VIEW.
	DIR=0
C						!ASSUME NO DIRECTION.
	IF(IABS(MLOC-HERE).NE.1) GO TO 200
C						!MIRROR TO N OR S?
	IF(MLOC.EQ.NRM) DIR=695
	IF(MLOC.EQ.SRM) DIR=699
C						!DIR=N/S.
	IF(MOD(MDIR,180).NE.0) GO TO 100
C						!MIRROR N-S?
	CALL RSPSUB(847,DIR)
C						!YES, HE SEES PANEL
	CALL RSPSB2(848,DIR,DIR)
C						!AND NARROW ROOMS.
	GO TO 200
C
100	M1=MRHERE(HERE)
C						!WHICH MIRROR?
	MRBF=0
C						!ASSUME INTACT.
	IF(((M1.EQ.1).AND..NOT.MR1F).OR.
     &	((M1.EQ.2).AND..NOT.MR2F)) MRBF=1
	CALL RSPSUB(849+MRBF,DIR)
C						!DESCRIBE.
	IF((M1.EQ.1).AND.MROPNF) CALL RSPEAK(823+MRBF)
	IF(MRBF.NE.0) CALL RSPEAK(851)
C
200	I=0
C						!ASSUME NO MORE TO DO.
	IF((NT.EQ.0).AND.((DIR.EQ.0).OR.(DIR.EQ.699))) I=852
	IF((ST.EQ.0).AND.((DIR.EQ.0).OR.(DIR.EQ.695))) I=853
	IF((NT+ST+DIR).EQ.0) I=854
	IF(HT.NE.0) CALL RSPEAK(I)
C						!DESCRIBE HALLS.
	RETURN
	END
C
C EWTELL--	DESCRIBE E/W NARROW ROOMS
	SUBROUTINE EWTELL(RM,ST)
	IMPLICIT INTEGER(A-Z)
	LOGICAL M1
C
C ROOMS
	include 'rindex.h'
	include 'flags.h'
C EWTELL, PAGE 2
C
C NOTE THAT WE ARE EAST OR WEST OF MIRROR, AND
C MIRROR MUST BE N-S.
C
	M1=(MDIR+(MOD(RM-MRAE,2)*180)).EQ.180
	I=819+MOD(RM-MRAE,2)
C						!GET BASIC E/W STRING.
	IF((M1.AND..NOT.MR1F).OR.(.NOT.M1.AND..NOT.MR2F))
     &	I=I+2
	CALL RSPEAK(I)
	IF(M1.AND.MROPNF) CALL RSPEAK(823+((I-819)/2))
	CALL RSPEAK(825)
	CALL RSPEAK(ST)
	RETURN
	END

File: C2021/nrooms.c
// Copyright (c) 1980, InfoCom Computers and Communications, Cambridge MA 02142
// All rights reserved, commercial usage strictly prohibited.
// Written by R. M. Supnik.
// Revisions Copyright (c) 2021, Darth Spectra (Lydia Marie Williamson).
#include "extern.h"
#include "common.h"

static void lookto(int, int, int, int, int);
static void ewtell(int, int);

// Special purpose room routines, part 2
Bool rappl2(int ri) {
// Initialized data
   const int newrms = 38;

// System generated locals
   int i__1;
   Bool ret_val;

// Local variables
   int i;
   int j;

// ROOMS

// RAPPL2, PAGE 2

   ret_val = true;
   switch (ri - newrms + 1) {
      case 1:
         goto L38000;
      case 2:
         goto L39000;
      case 3:
         goto L40000;
      case 4:
         goto L41000;
      case 5:
         goto L42000;
      case 6:
         goto L43000;
      case 7:
         goto L44000;
      case 8:
         goto L45000;
      case 9:
         goto L46000;
      case 10:
         goto L47000;
      case 11:
         goto L48000;
      case 12:
         goto L49000;
      case 13:
         goto L50000;
      case 14:
         goto L51000;
      case 15:
         goto L52000;
      case 16:
         goto L53000;
      case 17:
         goto L54000;
      case 18:
         goto L55000;
      case 19:
         goto L56000;
      case 20:
         goto L57000;
      case 21:
         goto L58000;
      case 22:
         goto L59000;
      case 23:
         goto L60000;
   }
   bug(70, ri);
   return ret_val;

// R38--	MIRROR D ROOM

L38000:
   if (prsvec_1.prsa == vindex_1.lookw) {
      lookto(rindex_1.fdoor, rindex_1.mrg, 0, 682, 681);
   }
   return ret_val;

// R39--	MIRROR G ROOM

L39000:
   if (prsvec_1.prsa == vindex_1.walkiw) {
      jigsup(685);
   }
   return ret_val;

// R40--	MIRROR C ROOM

L40000:
   if (prsvec_1.prsa == vindex_1.lookw) {
      lookto(rindex_1.mrg, rindex_1.mrb, 683, 0, 681);
   }
   return ret_val;

// R41--	MIRROR B ROOM

L41000:
   if (prsvec_1.prsa == vindex_1.lookw) {
      lookto(rindex_1.mrc, rindex_1.mra, 0, 0, 681);
   }
   return ret_val;

// R42--	MIRROR A ROOM

L42000:
   if (prsvec_1.prsa == vindex_1.lookw) {
      lookto(rindex_1.mrb, 0, 0, 684, 681);
   }
   return ret_val;
// RAPPL2, PAGE 3

// R43--	MIRROR C EAST/WEST

L43000:
   if (prsvec_1.prsa == vindex_1.lookw) {
      ewtell(play_1.here, 683);
   }
   return ret_val;

// R44--	MIRROR B EAST/WEST

L44000:
   if (prsvec_1.prsa == vindex_1.lookw) {
      ewtell(play_1.here, 686);
   }
   return ret_val;

// R45--	MIRROR A EAST/WEST

L45000:
   if (prsvec_1.prsa == vindex_1.lookw) {
      ewtell(play_1.here, 687);
   }
   return ret_val;

// R46--	INSIDE MIRROR

L46000:
   if (prsvec_1.prsa != vindex_1.lookw) {
      return ret_val;
   }
// 						!LOOK?
   rspeak(688);
// 						!DESCRIBE

// NOW DESCRIBE POLE STATE.

// CASES 1,2--	MDIR=270 & MLOC=MRB, POLE IS UP OR IN HOLE
// CASES 3,4--	MDIR=0 V MDIR=180, POLE IS UP OR IN CHANNEL
// CASE 5--	POLE IS UP

   i = 689;
// 						!ASSUME CASE 5.
   if (findex_1.mdir == 270 && findex_1.mloc == rindex_1.mrb) {
      i = min(findex_1.poleuf, 1) + 690;
   }
   if (findex_1.mdir % 180 == 0) {
      i = min(findex_1.poleuf, 1) + 692;
   }
   rspeak(i);
// 						!DESCRIBE POLE.
   i__1 = findex_1.mdir / 45 + 695;
   rspsub(694, i__1);
// 						!DESCRIBE ARROW.
   return ret_val;
// RAPPL2, PAGE 4

// R47--	MIRROR EYE ROOM

L47000:
   if (prsvec_1.prsa != vindex_1.lookw) {
      return ret_val;
   }
// 						!LOOK?
   i = 704;
// 						!ASSUME BEAM STOP.
   i__1 = objcts_1.olnt;
   for (j = 1; j <= i__1; ++j) {
      if (qhere(j, play_1.here) && j != oindex_1.rbeam) {
         goto L47200;
      }
// L47100:
   }
   i = 703;
L47200:
   rspsub(i, objcts_1.odesc2[j - 1]);
// 						!DESCRIBE BEAM.
   lookto(rindex_1.mra, 0, 0, 0, 0);
// 						!LOOK NORTH.
   return ret_val;

// R48--	INSIDE CRYPT

L48000:
   if (prsvec_1.prsa != vindex_1.lookw) {
      return ret_val;
   }
// 						!LOOK?
   i = 46;
// 						!CRYPT IS OPEN/CLOSED.
   if ((objcts_1.oflag2[oindex_1.tomb - 1] & OpenO) != 0) {
      i = 12;
   }
   rspsub(705, i);
   return ret_val;

// R49--	SOUTH CORRIDOR

L49000:
   if (prsvec_1.prsa != vindex_1.lookw) {
      return ret_val;
   }
// 						!LOOK?
   rspeak(706);
// 						!DESCRIBE.
   i = 46;
// 						!ODOOR IS OPEN/CLOSED.
   if ((objcts_1.oflag2[oindex_1.odoor - 1] & OpenO) != 0) {
      i = 12;
   }
   if (findex_1.lcell == 4) {
      rspsub(707, i);
   }
// 						!DESCRIBE ODOOR IF THERE.
   return ret_val;

// R50--	BEHIND DOOR

L50000:
   if (prsvec_1.prsa != vindex_1.walkiw) {
      goto L50100;
   }
// 						!WALK IN?
   cevent_1.cflag[cindex_1.cevfol - 1] = true;
// 						!MASTER FOLLOWS.
   cevent_1.ctick[cindex_1.cevfol - 1] = -1;
   return ret_val;

L50100:
   if (prsvec_1.prsa != vindex_1.lookw) {
      return ret_val;
   }
// 						!LOOK?
   i = 46;
// 						!QDOOR IS OPEN/CLOSED.
   if ((objcts_1.oflag2[oindex_1.qdoor - 1] & OpenO) != 0) {
      i = 12;
   }
   rspsub(708, i);
   return ret_val;
// RAPPL2, PAGE 5

// R51--	FRONT DOOR

L51000:
   if (prsvec_1.prsa == vindex_1.walkiw) {
      cevent_1.ctick[cindex_1.cevfol - 1] = 0;
   }
// 						!IF EXITS, KILL FOLLOW.
   if (prsvec_1.prsa != vindex_1.lookw) {
      return ret_val;
   }
// 						!LOOK?
   lookto(0, rindex_1.mrd, 709, 0, 0);
// 						!DESCRIBE SOUTH.
   i = 46;
// 						!PANEL IS OPEN/CLOSED.
   if (findex_1.inqstf) {
      i = 12;
   }
// 						!OPEN IF INQ STARTED.
   j = 46;
// 						!QDOOR IS OPEN/CLOSED.
   if ((objcts_1.oflag2[oindex_1.qdoor - 1] & OpenO) != 0) {
      j = 12;
   }
   rspsb2(710, i, j);
   return ret_val;

// R52--	NORTH CORRIDOR

L52000:
   if (prsvec_1.prsa != vindex_1.lookw) {
      return ret_val;
   }
// 						!LOOK?
   i = 46;
   if ((objcts_1.oflag2[oindex_1.cdoor - 1] & OpenO) != 0) {
      i = 12;
   }
// 						!CDOOR IS OPEN/CLOSED.
   rspsub(711, i);
   return ret_val;

// R53--	PARAPET

L53000:
   if (prsvec_1.prsa == vindex_1.lookw) {
      i__1 = findex_1.pnumb + 712;
      rspsub(712, i__1);
   }
   return ret_val;

// R54--	CELL

L54000:
   if (prsvec_1.prsa != vindex_1.lookw) {
      return ret_val;
   }
// 						!LOOK?
   i = 721;
// 						!CDOOR IS OPEN/CLOSED.
   if ((objcts_1.oflag2[oindex_1.cdoor - 1] & OpenO) != 0) {
      i = 722;
   }
   rspeak(i);
   i = 46;
// 						!ODOOR IS OPEN/CLOSED.
   if ((objcts_1.oflag2[oindex_1.odoor - 1] & OpenO) != 0) {
      i = 12;
   }
   if (findex_1.lcell == 4) {
      rspsub(723, i);
   }
// 						!DESCRIBE.
   return ret_val;

// R55--	PRISON CELL

L55000:
   if (prsvec_1.prsa == vindex_1.lookw) {
      rspeak(724);
   }
// 						!LOOK?
   return ret_val;

// R56--	NIRVANA CELL

L56000:
   if (prsvec_1.prsa != vindex_1.lookw) {
      return ret_val;
   }
// 						!LOOK?
   i = 46;
// 						!ODOOR IS OPEN/CLOSED.
   if ((objcts_1.oflag2[oindex_1.odoor - 1] & OpenO) != 0) {
      i = 12;
   }
   rspsub(725, i);
   return ret_val;
// RAPPL2, PAGE 6

// R57--	NIRVANA AND END OF GAME

L57000:
   if (prsvec_1.prsa != vindex_1.walkiw) {
      return ret_val;
   }
// 						!WALKIN?
   rspeak(726);
   score(false);
// moved to exit routine	CLOSE(DBCH)
   exit_();

// R58--	TOMB ROOM

L58000:
   if (prsvec_1.prsa != vindex_1.lookw) {
      return ret_val;
   }
// 						!LOOK?
   i = 46;
// 						!TOMB IS OPEN/CLOSED.
   if ((objcts_1.oflag2[oindex_1.tomb - 1] & OpenO) != 0) {
      i = 12;
   }
   rspsub(792, i);
   return ret_val;

// R59--	PUZZLE SIDE ROOM

L59000:
   if (prsvec_1.prsa != vindex_1.lookw) {
      return ret_val;
   }
// 						!LOOK?
   i = 861;
// 						!ASSUME DOOR CLOSED.
   if (findex_1.cpoutf) {
      i = 862;
   }
// 						!OPEN?
   rspeak(i);
// 						!DESCRIBE.
   return ret_val;

// R60--	PUZZLE ROOM

L60000:
   if (prsvec_1.prsa != vindex_1.lookw) {
      return ret_val;
   }
// 						!LOOK?
   if (findex_1.cpushf) {
      goto L60100;
   }
// 						!STARTED PUZZLE?
   rspeak(868);
// 						!NO, DESCRIBE.
   if ((objcts_1.oflag2[oindex_1.warni - 1] & TChO) != 0) {
      rspeak(869);
   }
   return ret_val;

L60100:
   cpinfo(880, findex_1.cphere);
// 						!DESCRIBE ROOM.
   return ret_val;
}

// Describe view in mirror hallway
static void lookto(int nrm, int srm, int nt, int st, int ht) {
// System generated locals
   int i__1;

// Local variables
   int i, m1, dir, mrbf;

// LOOKTO, PAGE 2

   rspeak(ht);
// 						!DESCRIBE HALL.
   rspeak(nt);
// 						!DESCRIBE NORTH VIEW.
   rspeak(st);
// 						!DESCRIBE SOUTH VIEW.
   dir = 0;
// 						!ASSUME NO DIRECTION.
   if ((i__1 = findex_1.mloc - play_1.here, abs(i__1)) != 1) {
      goto L200;
   }
// 						!MIRROR TO N OR S?
   if (findex_1.mloc == nrm) {
      dir = 695;
   }
   if (findex_1.mloc == srm) {
      dir = 699;
   }
// 						!DIR=N/S.
   if (findex_1.mdir % 180 != 0) {
      goto L100;
   }
// 						!MIRROR N-S?
   rspsub(847, dir);
// 						!YES, HE SEES PANEL
   rspsb2(848, dir, dir);
// 						!AND NARROW ROOMS.
   goto L200;

L100:
   m1 = mrhere(play_1.here);
// 						!WHICH MIRROR?
   mrbf = 0;
// 						!ASSUME INTACT.
   if (m1 == 1 && !findex_1.mr1f || m1 == 2 && !findex_1.mr2f) {
      mrbf = 1;
   }
   i__1 = mrbf + 849;
   rspsub(i__1, dir);
// 						!DESCRIBE.
   if (m1 == 1 && findex_1.mropnf) {
      i__1 = mrbf + 823;
      rspeak(i__1);
   }
   if (mrbf != 0) {
      rspeak(851);
   }

L200:
   i = 0;
// 						!ASSUME NO MORE TO DO.
   if (nt == 0 && (dir == 0 || dir == 699)) {
      i = 852;
   }
   if (st == 0 && (dir == 0 || dir == 695)) {
      i = 853;
   }
   if (nt + st + dir == 0) {
      i = 854;
   }
   if (ht != 0) {
      rspeak(i);
   }
// 						!DESCRIBE HALLS.
}

// Describe e/w narrow rooms
static void ewtell(int rm, int st) {
// System generated locals
   int i__1;

// Local variables
   int i;
   Bool m1;

// ROOMS

// EWTELL, PAGE 2

// NOTE THAT WE ARE EAST OR WEST OF MIRROR, AND
// MIRROR MUST BE N-S.

   m1 = findex_1.mdir + (rm - rindex_1.mrae) % 2 * 180 == 180;
   i = (rm - rindex_1.mrae) % 2 + 819;
// 						!GET BASIC E/W STRING.
   if (m1 && !findex_1.mr1f || !m1 && !findex_1.mr2f) {
      i += 2;
   }
   rspeak(i);
   if (m1 && findex_1.mropnf) {
      i__1 = (i - 819) / 2 + 823;
      rspeak(i__1);
   }
   rspeak(825);
   rspeak(st);
}
